### ISSUE TITLE
Support address-sanitizer coverage collector

### ISSUE THREAD

**Issue by SimonKagstrom** [SimonKagstrom/kcov#42]
The Google address sanitizer (as of LLVM ~january this year) supports collecting code coverage:

https://code.google.com/p/address-sanitizer/wiki/AsanCoverage

output is gcov-like, but can be directed to a directory. Kcov should be able to collect this data and report it via a custom engine/file-parser.

**Comment by SimonKagstrom** [comment]
A basic version has been implemented with 34b13b5. However, it still has to resolve basic blocks in order to be meaningful. Looks like more of a disassembler is upcoming after all :-)

**Comment by attekett** [comment]
Do you know if the basic block resolve is feasible to implement? 

Now that Google is running [oss-fuzz](https://github.com/google/oss-fuzz), [libFuzzer](http://llvm.org/docs/LibFuzzer.html) and [SanitizerCoverage](http://clang.llvm.org/docs/SanitizerCoverage.html) compile option is available for many projects.

SanitizerCoverage has its own "Experimental" [coverage report](http://clang.llvm.org/docs/SanitizerCoverage.html#coverage-reports-experimental) functionality, but it is very simple and doesn't include HTML report generation. It does have a JSON report generator, that can be displayed in browser via a python http-server implementation.(That might help with implementation in kcov.)

**Comment by SimonKagstrom** [comment]
As far as I know, the dwarf debug information doesn't contain basic block
info, so I think it would be required to disassemble the binary.

This is certainly feasible using libbfd, but it would be a fairly big
change.


Anyway, kcov works as long as you compile with debug information, even
without the sanitizer coverage options.

**Comment by SimonKagstrom** [comment]
Should sort of work with b7a6831 in the clang-basic-block-handling branch. However, I see some strange results, so probably I haven't quite understood the address sanitizer output after all.

**Comment by SimonKagstrom** [comment]
@attekett: You probably know more about the clang address sanitizer than I do. I have an issue with the implementation in the clang-basic-block-handling branch, that I think really should work now. I don't seem to understand how clang generates calls to __sanitizer_cov, to me it doesn't seem sufficient to gather info on all executed basic blocks.

I build with ```-fsanitize=address -fsanitize-coverage=edge``` (=bb is the same), and use the following snippet (excluding main etc):

```c
static int second(int a)
{
	printf("second\n");

	a = a + 1;
	a = a + 2;
	a = a + 3;

	return a + 2;
}
```

In my view, this should generate at least two basic blocks: One until the printf and one for the rest of the code. However, disassembling it you get

```asm
00000000004e49d0 <second>:
  4e49d0:       55                      push   %rbp
  4e49d1:       48 89 e5                mov    %rsp,%rbp
  4e49d4:       48 83 ec 10             sub    $0x10,%rsp
  4e49d8:       31 c0                   xor    %eax,%eax
  4e49da:       3b 05 a4 f3 e9 00       cmp    0xe9f3a4(%rip),%eax        # 1383d84 <_ZN7__ubsan11ubsan_flagsE+0x2c>
  4e49e0:       89 7d f8                mov    %edi,-0x8(%rbp)
  4e49e3:       0f 8c 13 00 00 00       jl     4e49fc <second+0x2c>
  4e49e9:       48 8d 05 80 f3 e9 00    lea    0xe9f380(%rip),%rax        # 1383d70 <_ZN7__ubsan11ubsan_flagsE+0x18>
  4e49f0:       48 83 c0 14             add    $0x14,%rax
  4e49f4:       48 89 c7                mov    %rax,%rdi
  4e49f7:       e8 34 26 ff ff          callq  4d7030 <__sanitizer_cov>
  4e49fc:       48 8d 3d 5d 44 01 00    lea    0x1445d(%rip),%rdi        # 4f8e60 <.str.1>
  4e4a03:       8b 45 f8                mov    -0x8(%rbp),%eax
  4e4a06:       89 45 fc                mov    %eax,-0x4(%rbp)
  4e4a09:       b0 00                   mov    $0x0,%al
  4e4a0b:       e8 50 13 f5 ff          callq  435d60 <__interceptor_printf>
  4e4a10:       8b 4d fc                mov    -0x4(%rbp),%ecx
  4e4a13:       83 c1 01                add    $0x1,%ecx
  4e4a16:       89 4d fc                mov    %ecx,-0x4(%rbp)
  4e4a19:       8b 4d fc                mov    -0x4(%rbp),%ecx
  4e4a1c:       83 c1 02                add    $0x2,%ecx
  4e4a1f:       89 4d fc                mov    %ecx,-0x4(%rbp)
  4e4a22:       8b 4d fc                mov    -0x4(%rbp),%ecx
  4e4a25:       83 c1 03                add    $0x3,%ecx
  4e4a28:       89 4d fc                mov    %ecx,-0x4(%rbp)
  4e4a2b:       8b 4d fc                mov    -0x4(%rbp),%ecx
  4e4a2e:       83 c1 02                add    $0x2,%ecx
  4e4a31:       89 45 f4                mov    %eax,-0xc(%rbp)
  4e4a34:       89 c8                   mov    %ecx,%eax
  4e4a36:       48 83 c4 10             add    $0x10,%rsp
  4e4a3a:       5d                      pop    %rbp
  4e4a3b:       c3                      retq   
  4e4a3c:       0f 1f 40 00             nopl   0x0(%rax)
```

so the first part (up to printf) is OK (apart from the ubsan_flags check, which generates extra basic blocks), but the instructions after are not headed with a call to __sanitizer_cov. Using the sancov tool with -html-report also misses the arithmetic instructions afterwards, which are useless of course, but part of the instruction stream, so it's not only kcov.

Any ideas?

**Comment by kcc** [comment]
Please note that we are moving towards deprecating what currently is `-fsanitize-coverage=edge`
and using `-fsanitize-coverage=trace-pc-guard` instead. 
See also https://clang.llvm.org/docs/SanitizerCoverage.html

But for your question this is not important. 
In LLVM a basic block may contain an arbitrary number of calls and so we will insert only one instrumentation point in such BB. 

For more questions, please contact us in https://github.com/google/sanitizers
or at llvm-dev@lists.llvm.org

**Comment by attekett** [comment]
@kcc Thanks for your quick reply. :)

**Comment by SimonKagstrom** [comment]
Thanks to both of you for the replies! I'll move further questions to the proper place, but I'd like to make one clarification.

For code coverage, I think this case really matters. Consider the following snippet:

```
void do_something()
{
    exit(0)
}

static int second(int a)
{
        a = 6;

	do_something();

	a = a + 1;
	a = a + 2;
	a = a + 3;

	return a + 2;
}
```

With only a single instrumentation point, there is no way of determining if the section after do_something() is actually executed, even if the instrumentation call in the first basic block triggers. Indeed, in this case, the lines won't execute.

I should say that I also tried the sanitizer-coverage-prune-blocks=false option after peeking in the source code, but it doesn't change my situation.

**Comment by kcc** [comment]
> I think this case really matters. Consider the following snippet:

It does matter, but you will have coverage signal from `do_something`.
This makes such coverage instrumentation suitable for fuzzing, but not necessary very good for visualization. I agree with that. 

> I also tried the sanitizer-coverage-prune-blocks=false option after peeking in the source code, but it doesn't change my situation.

That's mostly unrelated. It will avoid instrumenting some of the basic blocks in cases where 
these basic blocks will not add any extra information. 
Consider 
```
// A
if (...) {
  // B
} else {
  // C
}
// D
```

If we instrument B and C, there is no point in instrumenting A and D (for the sake of fuzzing).

**Comment by SimonKagstrom** [comment]
kcov support is basically fixed with fd1aac2 etc, but see Issue #184 for problems with tracing certain common constructs.

New issues will be opened for remaining problems/features.
